### PR TITLE
Add missing is_function guards to stdlib

### PR DESCRIPTION
**Pull request by gmile**

### CODE REVIEW

File: lib/elixir/lib/kernel/cli.ex [hunk 1]
```diff
@@ -67,7 +67,7 @@ defmodule Kernel.CLI do
     if hooks == [], do: res, else: at_exit(res)
   end
 
-  defp exec_fun(fun, res) when is_function(fun, 1) and is_tuple(res) do
```
Comment (gmile) [inline lib/elixir/lib/kernel/cli.ex]: Moving this one to public func, so it's closer to user.

File: lib/elixir/lib/gen_event/stream.ex [hunk 2]
```diff
@@ -56,7 +56,7 @@ defmodule GenEvent.Stream do
 end
 
 defimpl Enumerable, for: GenEvent.Stream do
-  def reduce(stream, acc, fun) do
+  def reduce(stream, acc, fun) when is_function(fun, 2) do
```
Comment (whatyouhide) [inline lib/elixir/lib/gen_event/stream.ex]: I don't think that this ever ends up getting called by the user directly, so maybe we can drop this one?

File: lib/elixir/lib/range.ex [hunk 3]
```diff
@@ -76,7 +76,7 @@ defmodule Range do
 end
 
 defimpl Enumerable, for: Range do
-  def reduce(first..last, acc, fun) do
```
Comment (whatyouhide) [inline lib/elixir/lib/range.ex]: Same as `GenEvent.Stream`.

**Comment by josevalim** [comment]
I am not sure we want this. I find the error message for calling a function with more or less arguments to be better than an UndefinedFunctionClause error.

File: lib/elixir/lib/kernel/cli.ex (hunk 1 above)
Comment (lexmag) [inline lib/elixir/lib/kernel/cli.ex]: With this change `at_exit` hooks won't be checked.

**Comment by josevalim** [comment]
Given my previous comment, I would like to propose the following: let's add guards to `Stream` only because those functions are stored and invoked later, so the error can be confusing. This will also make Stream on pair with Enum. I would avoid adding the other ones.

File: lib/elixir/lib/gen_event/stream.ex (hunk 2 above)
Comment (josevalim) [inline lib/elixir/lib/gen_event/stream.ex]: Yes, let's drop all checks happening inside protocol implementations.

File: lib/elixir/lib/kernel/cli.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/kernel/cli.ex]: This function is called recursively via at_exit so I would keep the guard here.

**Comment by josevalim** [comment]
Ignore me. Let's add this. :D

File: lib/elixir/lib/kernel/cli.ex (hunk 1 above)
Comment (gmile) [inline lib/elixir/lib/kernel/cli.ex]: Understood. I will revert this one.

**Comment by gmile** [comment]
@josevalim so to clarify – apart from adding guard on `exec_fun ` in `kernel/cli.ex`, no other changes to be made here? E.g. it's OK to keep newly added guards inside protocol implementations?

**Comment by whatyouhide** [comment]
@gmile no I think guards inside protocols should still go, it's the rest that's supposed to stay. Correct @josevalim?

**Comment by gmile** [comment]
@whatyouhide ah, I might have misunderstood then, sorry. Ok, let's wait for clarification from Jose

**Comment by josevalim** [comment]
@gmile ignore my comments on this main thread but please accept my comments in the code itself (to avoid confusion, I have removed the old comments).

**Comment by gmile** [comment]
Reverted changes related to protocols. How I didn't miss any

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
:heart:

**Pull request merged**